### DISCUSSION TITLE
Potential Bugs Report

### DISCUSSION THREAD

**Discussion by GANGE666** [libvips/libvips#2780] [General]
Hi, I recently scanned **libvips 8.12.2** using my static analysis tools. After manual analysis of the reports, I identified several potential vulnerabilities. Please check them and patch if necessary.

## **Divide Zero**

1. in function `vips_avg_build` , [avg.c#L103](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/arithmetic/avg.c#L103)

```c
vals = (gint64) 
	vips_image_get_width( statistic->in ) * 
	vips_image_get_height( statistic->in ) * 
	vips_image_get_bands( statistic->in );
average = avg->sum / vals;
```

The return value of `vips_image_get_width` and `vips_image_get_height` could be 0 in some case. 

**Root cause**: When initial vip image in function `vips_image_init_fields` ([header.c#L1057](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/iofuncs/header.c#L1057)), there is absent of checking the `width` and `height` of `image` . So that these values depend entirely on the caller's incoming parameters.

```c
void 
vips_image_init_fields( VipsImage *image, 
	int xsize, int ysize, int bands, 
	VipsBandFormat format, VipsCoding coding, 
	VipsInterpretation interpretation, 
	double xres, double yres )
{
	g_object_set( image,
		"width", xsize,
		"height", ysize,
		"bands", bands,
		"format", format,
		NULL );
```

However, not every caller has constraints on `xsize` and `ysize`(width & height). For example, in function `vips__analyze_read_header` ([analyze2vips.c#L524](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/foreign/analyze2vips.c#L524)), `width` and `height`  are parsed directly from the file content. When encountering a corrupt or malicious file, the `width` and `length` can be set to **arbitrary value**.

```c
int vips__analyze_read_header( const char *filename, VipsImage *out )
{
// ...
	generate_filenames( filename, header, image );
	if( !(d = read_header( header )) ) 
		return( -1 );

	if( get_vips_properties( d, &width, &height, &bands, &fmt ) ) {
		g_free( d );
		return( -1 );
	}

	vips_image_init_fields( out,
		width, height, bands, fmt, 
		VIPS_CODING_NONE, 
		bands == 1 ? 
			VIPS_INTERPRETATION_B_W : VIPS_INTERPRETATION_sRGB, 
		1.0, 1.0 );
// ...
}
```

1. in function `vips_deviate_build` , [deviate.c#L90](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/arithmetic/deviate.c#L90)

```c
vals = (gint64) 
	vips_image_get_width( statistic->in ) * 
	vips_image_get_height( statistic->in ) * 
	vips_image_get_bands( statistic->in );
s = deviate->sum;
s2 = deviate->sum2;

g_object_set( object, 
	"out", sqrt( VIPS_FABS( s2 - (s * s / vals) ) / (vals - 1) ),
	NULL );
```

The return value of `vips_image_get_width` and `vips_image_get_height` could be 0 in some case. 

Root cause as same as 1 above.

## Out-of-**Bound**

1. in function `vips_draw_smudge_build` , [draw_smudge.c#L85](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/draw/draw_smudge.c#L85)

```c
int bands = vips_image_get_bands( draw->image ) * 
	(vips_band_format_iscomplex( vips_image_get_format( im ) ) ? 
	 	2 : 1);
int elements = bands * vips_image_get_width( im );
// ...
#define SMUDGE( TYPE ) \
	for( y = 0; y < clipped.height; y++ ) { \
		TYPE *q; \
		TYPE *p; \
		\
		q = (TYPE *) VIPS_IMAGE_ADDR( im, \
			clipped.left, clipped.top + y ); \
		p = q - elements - bands; \
		for( x = 0; x < clipped.width; x++ ) { \
			TYPE *p1, *p2; \
 			\
			for( b = 0; b < bands; b++ ) \
				total[b] = 0.0; \
			\
			p1 = p; \
			for( i = 0; i < 3; i++ ) { \
				p2 = p1; \
				for( j = 0; j < 3; j++ ) \
					for( b = 0; b < bands; b++ ) \
						total[b] += *p2++; \                                              // Out-of-Bounds here
				\
				p1 += elements; \
			} \
 			\
			for( b = 0; b < bands; b++ ) \
				q[b] = (16 * (double) q[b] + total[b]) / 25.0; \
			\
			p += bands; \
			q += bands; \
		} \
	}
```

Because the return value of `vips_image_get_width` mentioned above can be set to arbitrary value, I think there is a possibility that the array will be out of bounds ([Line#151](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/draw/draw_smudge.c#L151)) due to the large width.

**Comment by jcupitt** [comment]
Hello @GANGE666, thank you for reporting these findings, but I think it's a false positive.

> When initial vip image in function vips_image_init_fields ([header.c#L1057](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/iofuncs/header.c#L1057)), there is absent of checking the width and height of image.

The checking is done by glib. The width property (for example) is declared like this:

```
        VIPS_ARG_INT( class, "width", 2,
                _( "Width" ),
                _( "Image width in pixels" ),
                VIPS_ARGUMENT_SET_ALWAYS,
                G_STRUCT_OFFSET( VipsImage, Xsize ),
                1, VIPS_MAX_COORD, 1 );
```

The last three numbers are the min, max and default. You can test like this:

```C
/* compile with:
 *      gcc -g -Wall bounds.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>

int
main()
{       
        VipsImage *image;

        VIPS_INIT("");

        image = vips_image_new();
        vips_image_init_fields( image, 0, 0, 0,
                VIPS_FORMAT_UCHAR,
                VIPS_CODING_NONE,
                VIPS_INTERPRETATION_B_W,
                1.0, 1.0 );

        printf( "width set to %d\n", image->Xsize );

        return 0;
}  
```

I see:

```
$ ./a.out 

(process:761512): GLib-GObject-WARNING **: 13:58:41.445: value "0" of type 'gint' is invalid or out of range for property 'width' of type 'gint'

(process:761512): GLib-GObject-WARNING **: 13:58:41.445: value "0" of type 'gint' is invalid or out of range for property 'height' of type 'gint'

(process:761512): GLib-GObject-WARNING **: 13:58:41.445: value "0" of type 'gint' is invalid or out of range for property 'bands' of type 'gint'
width set to 1
```

**Reply by GANGE666** [reply]
Thanks a lot for your answer.

In addition, do you think that the excessively large width may cause the array `p2` to go out of bounds?
in [draw_smudge.c#L151](https://github.com/libvips/libvips/blob/43db3985d0c0f0321434f2bd1cda36c7dd4fa1e5/libvips/draw/draw_smudge.c#L151), as mentioned in #Out-of-Bound .1

**Reply by jcupitt** [reply]
I think that's OK -- `clipped` is within the image, and `p` is a pointer to the start of the line.

**Reply by GANGE666** [reply]
Thanks again. I have no more questions.
